### PR TITLE
[WIP] ConnectionBuilder and  FieldBuilder rework [DRAFT]

### PR DESCRIPTION
**Pull request by sungam3r**
Some ideas related to #2427

### CODE REVIEW

File: src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt [hunk 1]
```diff
@@ -51,11 +51,7 @@ namespace GraphQL.MicrosoftDI
         public GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> ResolveAsync(System.Func<GraphQL.IResolveFieldContext<TSourceType>, T1, T2, T3, T4, T5, System.Threading.Tasks.Task<TReturnType>> resolver) { }
         public GraphQL.MicrosoftDI.ResolverBuilder<TSourceType, TReturnType, T1, T2, T3, T4, T5> WithScope() { }
     }
-    public class ScopedAsyncFieldResolver<TReturnType> : GraphQL.Resolvers.AsyncFieldResolver<TReturnType>
-    {
-        public ScopedAsyncFieldResolver(System.Func<GraphQL.IResolveFieldContext, System.Threading.Tasks.Task<TReturnType>> resolver) { }
-    }
-    public class ScopedAsyncFieldResolver<TSourceType, TReturnType> : GraphQL.Resolvers.AsyncFieldResolver<TReturnType>
+    public class ScopedAsyncFieldResolver<TSourceType, TReturnType> : GraphQL.Resolvers.AsyncFieldResolver<TSourceType, TReturnType>
```
Comment (Shane32) [inline src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt:54]: I designed this to be flexible and reduce allocations.  With this design there is always an extra allocation required.  I will need to review my thought process and use cases.  Same thought applies to `ScopedFieldResolver`.

File: src/GraphQL.ApiTests/GraphQL.approved.txt [hunk 2]
```diff
@@ -501,8 +490,11 @@ namespace GraphQL.Builders
         public virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Directive(string name, System.Action<GraphQL.Types.AppliedDirective> configure) { }
         public virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Directive(string name, string argumentName, object argumentValue) { }
         public virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Name(string name) { }
-        public virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Resolve(GraphQL.Resolvers.IFieldResolver resolver) { }
+        public virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Resolve(GraphQL.Resolvers.IFieldResolver<TReturnType> resolver) { }
         public virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Resolve(System.Func<GraphQL.IResolveFieldContext<TSourceType>, TReturnType> resolve) { }
+        public virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> ResolveAsync(GraphQL.Resolvers.IFieldResolver<GraphQL.DataLoader.IDataLoaderResult<TReturnType>> resolver) { }
```
Comment (Shane32) [inline src/GraphQL.ApiTests/GraphQL.approved.txt:495]: It is important not to have `Async` in the method name here.  An asynchronous method returning a `Task<IDataLoaderResult>` is distinctly different than a synchronous method returning a `IDataLoaderResult`.  This is one of the bugs that people had been complaining about that was fixed in the new data loader design.  See #945 .

File: src/GraphQL.ApiTests/GraphQL.approved.txt [hunk 3]
```diff
@@ -441,28 +441,22 @@ namespace GraphQL
...
     public class ConnectionBuilder<TSourceType>
     {
         public GraphQL.Types.FieldType FieldType { get; set; }
-        public GraphQL.Builders.ConnectionBuilder<TSourceType> Argument<TArgumentGraphType>(string name, string description)
+        public GraphQL.Builders.ConnectionBuilder<TSourceType> Argument<TArgumentGraphType>(string name, System.Action<GraphQL.Types.QueryArgument> configure = null)
+            where TArgumentGraphType : GraphQL.Types.IGraphType { }
+        public GraphQL.Builders.ConnectionBuilder<TSourceType> Argument<TArgumentGraphType>(string name, string description, System.Action<GraphQL.Types.QueryArgument> configure = null)
             where TArgumentGraphType : GraphQL.Types.IGraphType { }
-        public GraphQL.Builders.ConnectionBuilder<TSourceType> Argument<TArgumentGraphType, TArgumentType>(string name, string description, TArgumentType defaultValue = default)
+        public GraphQL.Builders.ConnectionBuilder<TSourceType> Argument<TArgumentGraphType, TArgumentType>(string name, string description, TArgumentType defaultValue = default, System.Action<GraphQL.Types.QueryArgument> configure = null)
```
Comment (Shane32) [inline src/GraphQL.ApiTests/GraphQL.approved.txt:451]: This is good so long as there is no issue with the compiler questioning which overload to use.

File: src/GraphQL.ApiTests/GraphQL.approved.txt [hunk 4]
```diff
@@ -479,11 +473,6 @@ namespace GraphQL.Builders
             where TEdgeType : GraphQL.Types.Relay.EdgeType<TNodeType>
             where TConnectionType : GraphQL.Types.Relay.ConnectionType<TNodeType, TEdgeType> { }
     }
-    public static class FieldBuilder
-    {
-        public static GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Create<TSourceType, TReturnType>(GraphQL.Types.IGraphType type) { }
-        public static GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Create<TSourceType, TReturnType>(System.Type type = null) { }
-    }
```
Comment (Shane32) [inline src/GraphQL.ApiTests/GraphQL.approved.txt:486]: I understand that this "duplicates existing code", but I like `FieldBuilder.Create<>` better than `FieldBuilder<>.Create`.  Not a big deal though.

**Review by Shane32 — COMMENTED** [review]
I don't have time for a full review today, but I did review the api diff and made a few comments.

I don't think any breaking changes should be made for v4, but we can make any changes that are not breaking, and breaking changes can go in a PR for v5.

File: src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt:54]: Does the same apply to connection builder? I'm ok either way but both builders should act the same IMO - with or without TSourceType.

File: src/GraphQL.ApiTests/GraphQL.approved.txt (hunk 4 above)
Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:486]: One static method calls another one. It seems pointless to me.

File: src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt (hunk 1 above)
Comment (Shane32) [reply src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt:54]: Yes if it was public, although I had made it internal for reasons previously stated (and then eliminated redundant code). We can make it public if we expose bidirectional and similar as `get;` properties on the builder. This also eliminates an allocation.

Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt:54]: By the way, as for builders I'm not so concerned about allocations.

Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt:54]: And regarding these 3 properties in ConnectionBuilder. In fact, these are calculated properties - `_isUnidirectional` and `_isBidirectional`. The real object that is configured by builder is `FieldType`. `_pageSize` is not calculated but could be stored in `FieldType.Metadata` so `ConnectionBuilder` will store only reference to `FieldType` as `FieldBuilder` does. `IsUnidirectional` and `IsBidirectional` can be implemented in such a way that they set the arguments only if necessary.

Comment (Shane32) [reply src/GraphQL.ApiTests/GraphQL.MicrosoftDI.approved.txt:54]: The allocations I was referring to occur during execution. I am not counting allocations during build time.

**Pull request closed without merging**